Frontends/glslang/GlslangToTopVisitor.cpp

```diff
@@ -1391,7 +1391,12 @@ bool TGlslangToTopTraverser::visitAggregate(glslang::TVisit visit, glslang::TInt
         result = createUnaryIntrinsic(node->getOp(), precision, operands.front());
         break;
     default:
-        result = createIntrinsic(node->getOp(), precision, uniform, operands, glslangOperands.front()->getAsTyped()->getBasicType() == glslang::EbtUint);
+        // Check first for intrinsics that can be done natively
+        if (node->getOp() == glslang::EOpMix && 
+              gla::GetBasicTypeID(operands.front()) == llvm::Type::IntegerTyID)
+            result = llvmBuilder.CreateSelect(operands[2], operands[1], operands[0]);
```

Comment:
Hi John,

GLSL and GLES appear to disallow mixing of vector and scalar, at least for
the mix (and similar) intrinsics. Both require all arguments and result to
have the same number of components.

Let me know if you still have a concern with smearing.

Thanks,
Greg


Greg Fischer

Project Manager and Senior Engineer

LunarG, Inc. - 3D Driver Innovations

greg@lunarg.com <alan@lunarg.com>

On Thu, Jul 28, 2016 at 7:00 PM, John Kessenich <notifications@github.com>
wrote:

> In Frontends/glslang/GlslangToTopVisitor.cpp
> <https://github.com/LunarG/LunarGLASS/pull/31#discussion_r72728905>:
>
> > @@ -1391,7 +1391,12 @@ bool TGlslangToTopTraverser::visitAggregate(glslang::TVisit visit, glslang::TInt
> >          result = createUnaryIntrinsic(node->getOp(), precision, operands.front());
> >          break;
> >      default:
> > -        result = createIntrinsic(node->getOp(), precision, uniform, operands, glslangOperands.front()->getAsTyped()->getBasicType() == glslang::EbtUint);
> > +        // Check first for intrinsics that can be done natively
> > +        if (node->getOp() == glslang::EOpMix &&
> > +              gla::GetBasicTypeID(operands.front()) == llvm::Type::IntegerTyID)
> > +            result = llvmBuilder.CreateSelect(operands[2], operands[1], operands[0]);
>
> Are these all properly smeared or other to handle GLSL mixing of vectors
> and scalars?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LunarG/LunarGLASS/pull/31/files/e1b1fecbb59ce20f1e66ec70132e712953ac558b#r72728905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHL6q3gK63KCXmCOggKLqtdL3HW65CPOks5qaVCwgaJpZM4JXufM>
> .
>